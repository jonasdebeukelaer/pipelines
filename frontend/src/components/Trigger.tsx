/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Fab from '@material-ui/core/Fab';
import Checkbox from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import MenuItem from '@material-ui/core/MenuItem';
import * as React from 'react';
import { stylesheet } from 'typestyle';
import { ApiTrigger } from '../apis/job';
import { HelpButton } from '../atoms/HelpButton';
import Input from '../atoms/Input';
import Separator from '../atoms/Separator';
import { commonCss, padding } from '../Css';
import {
  buildCron,
  buildTrigger,
  dateToPickerFormat,
  getIntervalPeriodAndValue,
  PeriodicInterval,
  pickersToDate,
  triggers,
  TriggerType,
} from '../lib/TriggerUtils';

interface TriggerProps {
  trigger?: ApiTrigger;
  maxConcurrentRuns?: string;
  catchup?: boolean;
  onChange?: (config: {
    trigger?: ApiTrigger;
    maxConcurrentRuns?: string;
    catchup: boolean;
  }) => void;
}

interface TriggerState {
  cron: string;
  editCron: boolean;
  endDate: string;
  endTime: string;
  hasEndDate: boolean;
  hasStartDate: boolean;
  intervalCategory: PeriodicInterval;
  intervalValue: number;
  maxConcurrentRuns: string;
  selectedDays: boolean[];
  startDate: string;
  startTime: string;
  type: TriggerType;
  catchup: boolean;
}

const css = stylesheet({
  noMargin: {
    margin: 0,
  },
  marginRight: {
    margin: '0 5px 0 0',
  },
});

export default class Trigger extends React.Component<TriggerProps, TriggerState> {
  constructor(props: TriggerProps) {
    super(props);

    let startDateTime: Date;
    if (!!props.trigger?.cron_schedule?.start_time) {
      startDateTime = new Date(props.trigger.cron_schedule.start_time);
    } else if (!!props.trigger?.periodic_schedule?.start_time) {
      startDateTime = new Date(props.trigger.periodic_schedule.start_time);
    } else {
      startDateTime = new Date();
    }

    let endDateTime: Date;
    if (!!props.trigger?.cron_schedule?.end_time) {
      endDateTime = new Date(props.trigger.cron_schedule.end_time);
    } else if (!!props.trigger?.periodic_schedule?.end_time) {
      endDateTime = new Date(props.trigger.periodic_schedule.end_time);
    } else {
      endDateTime = new Date(startDateTime);
      // default to a week after the start date
      endDateTime.setDate(endDateTime.getDate() + 7);
    }

    const [startDate, startTime] = dateToPickerFormat(startDateTime);
    const [endDate, endTime] = dateToPickerFormat(endDateTime);

    let intervalCategory = PeriodicInterval.MINUTE;
    let intervalValue = 1;
    if (!!props.trigger?.periodic_schedule?.interval_second) {
      [intervalCategory, intervalValue] = getIntervalPeriodAndValue(
        props.trigger.periodic_schedule.interval_second,
      );
    }

    let selectedDays;
    if (intervalCategory === PeriodicInterval.WEEK && !!props.trigger?.cron_schedule?.cron) {
      selectedDays = new Array(7).fill(false);
      const days = props.trigger.cron_schedule.cron.split(' ')[5].split(',');
      days.every(day_str => (selectedDays[parseInt(day_str)] = true));
    } else {
      selectedDays = new Array(7).fill(true);
    }

    this.state = {
      catchup: props.catchup === undefined ? true : props.catchup,
      cron: !!props.trigger?.cron_schedule?.cron ? props.trigger.cron_schedule.cron : '0 * * * * ?',
      editCron: !!props.trigger?.cron_schedule?.cron,
      endDate: endDate,
      endTime: endTime,
      hasEndDate:
        !!props.trigger?.cron_schedule?.end_time || !!props.trigger?.periodic_schedule?.end_time,
      hasStartDate:
        !!props.trigger?.cron_schedule?.start_time ||
        !!props.trigger?.periodic_schedule?.start_time,
      intervalCategory: intervalCategory,
      intervalValue: intervalValue,
      maxConcurrentRuns: props.maxConcurrentRuns ? props.maxConcurrentRuns : '10',
      selectedDays: selectedDays,
      startDate: startDate,
      startTime: startTime,
      type: !!props.trigger?.cron_schedule ? TriggerType.CRON : TriggerType.INTERVALED,
    };
  }

  public render(): JSX.Element {
    const {
      cron,
      editCron,
      endDate,
      endTime,
      hasEndDate,
      hasStartDate,
      intervalCategory,
      intervalValue,
      maxConcurrentRuns,
      selectedDays,
      startDate,
      startTime,
      type,
      catchup,
    } = this.state;

    return (
      <div>
        <Input
          select={true}
          label='Trigger type'
          required={true}
          onChange={this.handleChange('type')}
          value={type}
          variant='outlined'
        >
          {Array.from(triggers.entries()).map((trigger, i) => (
            <MenuItem key={i} value={trigger[0]}>
              {trigger[1].displayName}
            </MenuItem>
          ))}
        </Input>

        <div>
          <Input
            label='Maximum concurrent runs'
            required={true}
            type='number'
            onChange={this.handleChange('maxConcurrentRuns')}
            value={maxConcurrentRuns}
            error={+maxConcurrentRuns > 10 || +maxConcurrentRuns < 1}
            variant='outlined'
          />

          <div className={commonCss.flex}>
            <FormControlLabel
              className={padding(7, 't')}
              control={
                <Checkbox
                  checked={hasStartDate}
                  color='primary'
                  onClick={this.handleChange('hasStartDate')}
                />
              }
              label='Has start date'
            />
            <Input
              label='Start date'
              type='date'
              onChange={this.handleChange('startDate')}
              value={startDate}
              width={160}
              variant='outlined'
              InputLabelProps={{ classes: { outlined: css.noMargin }, shrink: true }}
              style={{ visibility: hasStartDate ? 'visible' : 'hidden' }}
            />
            <Separator />
            <Input
              label='Start time'
              type='time'
              onChange={this.handleChange('startTime')}
              value={startTime}
              width={120}
              variant='outlined'
              InputLabelProps={{ classes: { outlined: css.noMargin }, shrink: true }}
              style={{ visibility: hasStartDate ? 'visible' : 'hidden' }}
            />
          </div>

          <div className={commonCss.flex}>
            <FormControlLabel
              className={padding(7, 't')}
              control={
                <Checkbox
                  checked={hasEndDate}
                  color='primary'
                  onClick={this.handleChange('hasEndDate')}
                />
              }
              label='Has end date'
            />
            <Input
              label='End date'
              type='date'
              onChange={this.handleChange('endDate')}
              value={endDate}
              width={160}
              style={{ visibility: hasEndDate ? 'visible' : 'hidden' }}
              InputLabelProps={{ classes: { outlined: css.noMargin }, shrink: true }}
              variant='outlined'
            />
            <Separator />
            <Input
              label='End time'
              type='time'
              onChange={this.handleChange('endTime')}
              value={endTime}
              width={120}
              style={{ visibility: hasEndDate ? 'visible' : 'hidden' }}
              InputLabelProps={{ classes: { outlined: css.noMargin }, shrink: true }}
              variant='outlined'
            />
          </div>
          <div className={commonCss.flex}>
            <FormControlLabel
              control={
                <Checkbox
                  checked={catchup}
                  color='primary'
                  onClick={this.handleChange('catchup')}
                />
              }
              label='Catchup'
            />
            <HelpButton
              helpText={
                <div>
                  <p>
                    Whether the recurring run should catch up if behind schedule. Defaults to true.
                  </p>
                  <p>
                    For example, if the recurring run is paused for a while and re-enabled
                    afterwards. If catchup=true, the scheduler will catch up on (backfill) each
                    missed interval. Otherwise, it only schedules the latest interval if more than
                    one interval is ready to be scheduled.
                  </p>
                  <p>
                    Usually, if your pipeline handles backfill internally, you should turn catchup
                    off to avoid duplicate backfill.
                  </p>
                </div>
              }
            />
          </div>

          <div className={commonCss.flex}>
            <span className={padding(15, 't')}>Run every</span>
            {type === TriggerType.INTERVALED && (
              <div className={commonCss.flex}>
                <Separator />
                <Input
                  required={true}
                  type='number'
                  onChange={this.handleChange('intervalValue')}
                  value={intervalValue}
                  height={30}
                  width={65}
                  error={intervalValue < 1}
                  variant='outlined'
                />
              </div>
            )}
            <Separator />
            <Input
              id='intervalCategory'
              required={true}
              select={true}
              onChange={this.handleChange('intervalCategory')}
              value={intervalCategory}
              height={30}
              width={95}
              variant='outlined'
              disabled={type === TriggerType.CRON && editCron}
            >
              {Object.keys(PeriodicInterval).map((interval, i) => (
                <MenuItem key={i} value={PeriodicInterval[interval]}>
                  {PeriodicInterval[interval] + (type === TriggerType.INTERVALED ? 's' : '')}
                </MenuItem>
              ))}
            </Input>

            {type === TriggerType.CRON && intervalCategory === PeriodicInterval.WEEK && (
              <div className={padding(16, 'lt')}>
                <span>On:</span>
                <FormControlLabel
                  control={
                    <Checkbox
                      checked={this._isAllDaysChecked()}
                      color='primary'
                      onClick={this._toggleCheckAllDays.bind(this)}
                      disabled={type === TriggerType.CRON && editCron}
                    />
                  }
                  label='All'
                />
                <Separator />
                {['S', 'M', 'T', 'W', 'T', 'F', 'S'].map((day, i) => (
                  <Fab
                    className={css.marginRight}
                    size='small'
                    key={i}
                    onClick={() => this._toggleDay(i)}
                    color={selectedDays[i] ? 'primary' : 'secondary'}
                    disabled={type === TriggerType.CRON && editCron}
                  >
                    {day}
                  </Fab>
                ))}
              </div>
            )}
          </div>
        </div>

        {type === TriggerType.CRON && (
          <div>
            <div className={commonCss.flex}>
              <FormControlLabel
                control={
                  <Checkbox
                    checked={editCron}
                    color='primary'
                    onClick={this.handleChange('editCron')}
                  />
                }
                label={
                  <span>
                    Allow editing cron expression. ( format is specified{' '}
                    <a href='https://godoc.org/github.com/robfig/cron#hdr-CRON_Expression_Format'>
                      here
                    </a>
                    )
                  </span>
                }
              />
            </div>

            <Input
              label='cron expression'
              onChange={this.handleChange('cron')}
              value={cron}
              width={300}
              disabled={!editCron}
              error={editCron && cron.split(' ').length !== 6}
              variant='outlined'
            />

            <div>Note: Start and end dates/times are handled outside of cron.</div>
          </div>
        )}
      </div>
    );
  }

  public handleChange = (name: string) => (event: any) => {
    const target = event.target;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    // Make sure the desired field is set on the state object first, then
    // use the state values to compute the new trigger
    this.setState(
      {
        [name]: value,
      } as any,
      this._updateTrigger,
    );
  };

  private _updateTrigger = () => {
    const {
      hasStartDate,
      hasEndDate,
      startDate,
      startTime,
      endDate,
      endTime,
      editCron,
      intervalCategory,
      intervalValue,
      type,
      cron,
      selectedDays,
      catchup,
    } = this.state;

    const startDateTime = pickersToDate(hasStartDate, startDate, startTime);
    const endDateTime = pickersToDate(hasEndDate, endDate, endTime);

    // TODO: Why build the cron string unless the TriggerType is not CRON?
    // Unless cron editing is enabled, calculate the new cron string, set it in state,
    // then use it to build new trigger object and notify the parent
    this.setState(
      {
        cron: editCron ? cron : buildCron(startDateTime, intervalCategory, selectedDays),
      },
      () => {
        const trigger = buildTrigger(
          intervalCategory,
          intervalValue,
          startDateTime,
          endDateTime,
          type,
          this.state.cron,
        );

        if (this.props.onChange) {
          this.props.onChange({
            catchup,
            maxConcurrentRuns: trigger ? this.state.maxConcurrentRuns : undefined,
            trigger,
          });
        }
      },
    );
  };

  private _isAllDaysChecked(): boolean {
    return this.state.selectedDays.every(d => !!d);
  }

  private _toggleCheckAllDays(): void {
    const isAllChecked = this._isAllDaysChecked();
    this.state.selectedDays.forEach((d, i) => {
      if (d !== !isAllChecked) {
        this._toggleDay(i);
      }
    });
  }

  private _toggleDay(index: number): void {
    const newDays = this.state.selectedDays;
    newDays[index] = !newDays[index];
    const startDate = pickersToDate(
      this.state.hasStartDate,
      this.state.startDate,
      this.state.startTime,
    );
    const cron = buildCron(startDate, this.state.intervalCategory, this.state.selectedDays);

    this.setState(
      {
        cron,
        selectedDays: newDays,
      },
      this._updateTrigger,
    );
  }
}
