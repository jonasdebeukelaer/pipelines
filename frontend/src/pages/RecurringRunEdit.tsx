/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Buttons from '../lib/Buttons';
import RunUtils from '../lib/RunUtils';
import { ApiExperiment } from '../apis/experiment';
import { ApiJob, ApiTrigger } from '../apis/job';
import { Apis } from '../lib/Apis';
import { Page } from './Page';
import { RoutePage, RouteParams } from '../components/Router';
import { Breadcrumb, ToolbarProps } from '../components/Toolbar';
import { classes } from 'typestyle';
import { commonCss, padding } from '../Css';
import { errorToMessage, logger } from '../lib/Utils';
import Input from '../atoms/Input';
import { TextFieldProps } from '@material-ui/core/TextField';
import Trigger from '../components/Trigger';
import BusyButton from '../atoms/BusyButton';
import Button from '@material-ui/core/Button';
import RunParameters from '../components/RunParameters';
import { ApiParameter } from '../apis/pipeline';

interface EditRecurringRunState {
  run?: ApiJob;
  runName: string;
  description: string;
  serviceAccount: string;
  trigger?: ApiTrigger;
  maxConcurrentRuns: string;
  catchup: boolean;
  parameters: ApiParameter[];
  errorMessage: string;
  isSaving: boolean;
}

class RecurringRunEdit extends Page<{}, EditRecurringRunState> {
  public state: EditRecurringRunState = {
    runName: '',
    description: '',
    serviceAccount: '',
    maxConcurrentRuns: '',
    catchup: true,
    parameters: [],
    errorMessage: '',
    isSaving: false,
  };

  public getInitialToolbarState(): ToolbarProps {
    const buttons = new Buttons(this.props, this.refresh.bind(this));
    return {
      actions: buttons
        .refresh(this.refresh.bind(this))
        .delete(
          () => (this.state.run ? [this.state.run!.id!] : []),
          'recurring run config',
          this._deleteCallback.bind(this),
          true /* useCurrentResource */,
        )
        .getToolbarActionMap(),
      breadcrumbs: [],
      pageTitle: '',
    };
  }

  public render(): JSX.Element {
    const {
      run,
      runName,
      description,
      serviceAccount,
      trigger,
      maxConcurrentRuns,
      catchup,
      parameters,
      errorMessage,
      isSaving,
    } = this.state;

    return (
      <div className={classes(commonCss.page, padding(20, 'lr'))}>
        {!!run && (
          <div className={commonCss.page}>
            <div className={padding(10, 't')}>
              <div className={commonCss.header}>General Params</div>

              <Input
                id='runName'
                label='Run name'
                required={true}
                multiline={false}
                onChange={this._handleChange('runName')}
                value={runName}
                error={runName === ''}
                variant='outlined'
              />

              <Input
                id='description'
                label='Description'
                multiline={false}
                onChange={this._handleChange('description')}
                value={description}
                variant='outlined'
              />

              <Input
                id='serviceAccount'
                label='Service Account'
                multiline={false}
                onChange={this._handleChange('serviceAccount')}
                value={serviceAccount}
                variant='outlined'
              />

              <div className={commonCss.header}>Run trigger</div>

              <Trigger
                trigger={trigger}
                maxConcurrentRuns={maxConcurrentRuns}
                catchup={catchup || true}
                onChange={({ trigger, maxConcurrentRuns, catchup }) =>
                  this._updateRunTriggers(trigger, maxConcurrentRuns, catchup)
                }
              />

              {!!parameters && (
                <RunParameters
                  initialParams={parameters}
                  titleMessage={
                    parameters.length > 0
                      ? 'Edit pipeline parameters'
                      : 'This pipeline has no parameters'
                  }
                  handleParamChange={this._handleParamChange.bind(this)}
                />
              )}
            </div>

            <div className={classes(commonCss.flex, padding(40, 'tb'))}>
              <BusyButton
                id='saveBtn'
                disabled={!!errorMessage}
                busy={isSaving}
                className={commonCss.buttonAction}
                title='Save'
                onClick={this._save.bind(this)}
              />
              <Button id='cancelBtn' onClick={this._goToRecurringDetailsPage.bind(this)}>
                Cancel
              </Button>

              <div id='errorMessage' className={classes(padding(20, 'r'))} style={{ color: 'red' }}>
                {errorMessage}
              </div>

              {this._areParametersMissing() && (
                <div id='missingParametersMessage' style={{ color: 'orange' }}>
                  Some parameters are missing values
                </div>
              )}
            </div>
          </div>
        )}
      </div>
    );
  }

  private _handleChange = (name: string) => (event: any) => {
    const value = (event.target as TextFieldProps).value;
    this.setStateSafe({ [name]: value } as any, () => {
      this._validate();
    });
  };

  private _updateRunTriggers(
    trigger: ApiTrigger | undefined,
    maxConcurrentRuns: string | undefined,
    catchup: boolean,
  ) {
    this.setStateSafe(
      {
        trigger,
        maxConcurrentRuns,
        catchup,
      },
      () => this._validate(),
    );
  }

  private _handleParamChange(index: number, value: string): void {
    const { parameters } = this.state;
    if (!!parameters && parameters.length > index) {
      parameters[index].value = value;
      this.setStateSafe({ parameters });
    }
  }

  private _validate(): void {
    const { maxConcurrentRuns, runName, trigger } = this.state;
    try {
      if (!runName) {
        throw new Error('Run name is required');
      }

      RunUtils.ensureRecurringRunParamsAreValid(trigger, maxConcurrentRuns);

      this.setStateSafe({ errorMessage: '' });
    } catch (err) {
      this.setStateSafe({ errorMessage: err.message });
    }
  }

  public componentDidMount(): Promise<void> {
    return this.load();
  }

  public async refresh(): Promise<void> {
    await this.load();
  }

  public async load(): Promise<void> {
    this.clearBanner();
    const runId = this.props.match.params[RouteParams.runId];

    let run: ApiJob;
    try {
      run = await Apis.jobServiceApi.getJob(runId);
    } catch (err) {
      const errorMessage = await errorToMessage(err);
      await this.showPageError(
        `Error: failed to retrieve recurring run: ${runId}.`,
        new Error(errorMessage),
      );
      return;
    }

    const relatedExperimentId = RunUtils.getFirstExperimentReferenceId(run);
    let experiment: ApiExperiment | undefined;
    if (relatedExperimentId) {
      try {
        experiment = await Apis.experimentServiceApi.getExperiment(relatedExperimentId);
      } catch (err) {
        const errorMessage = await errorToMessage(err);
        await this.showPageError(
          `Error: failed to retrieve this recurring run's experiment.`,
          new Error(errorMessage),
          'warning',
        );
      }
    }
    const breadcrumbs: Breadcrumb[] = [];
    if (experiment) {
      breadcrumbs.push(
        { displayName: 'Experiments', href: RoutePage.EXPERIMENTS },
        {
          displayName: experiment.name!,
          href: RoutePage.EXPERIMENT_DETAILS.replace(
            ':' + RouteParams.experimentId,
            experiment.id!,
          ),
        },
      );
    } else {
      breadcrumbs.push({ displayName: 'All runs', href: RoutePage.RUNS });
    }
    const pageTitle = 'Edit recurring run';

    this.props.updateToolbar({ breadcrumbs, pageTitle });

    this.setState({
      run,
      runName: run.name || '',
      description: run.description || '',
      serviceAccount: run.service_account || '',
      trigger: run.trigger,
      maxConcurrentRuns: run.max_concurrency || '',
      catchup: !run.no_catchup,
      parameters: run.pipeline_spec?.parameters || [],
    });
  }

  private _deleteCallback(_: string[], success: boolean): void {
    if (success) {
      const breadcrumbs = this.props.toolbarProps.breadcrumbs;
      const previousPage = breadcrumbs.length
        ? breadcrumbs[breadcrumbs.length - 1].href
        : RoutePage.EXPERIMENTS;
      this.props.history.push(previousPage);
    }
  }

  private _areParametersMissing(): boolean {
    const { parameters } = this.state;
    return parameters?.some(p => !p.value);
  }

  private _goToRecurringDetailsPage(): void {
    const { run } = this.state;
    this.props.history.push(
      !!run?.id ? RoutePage.RECURRING_RUN.replace(':' + RouteParams.runId, run.id) : RoutePage.RUNS,
    );
  }

  private _save(): void {
    this.setStateSafe({ isSaving: true }, async () => {
      try {
        let updatedRun = this.state.run;

        if (!updatedRun || !updatedRun.pipeline_spec) {
          throw new Error('run object is not properly undefined');
        }

        updatedRun.name = this.state.runName;
        updatedRun.description = this.state.description;
        updatedRun.trigger = this.state.trigger;
        updatedRun.max_concurrency = this.state.maxConcurrentRuns;
        updatedRun.no_catchup = !this.state.catchup;
        updatedRun.pipeline_spec.parameters = this.state.parameters;

        await Apis.jobServiceApi.updateJob(updatedRun);

        this._goToRecurringDetailsPage();

        // TODO: snackbars are not working (trying to set state (in RoutePage) whilst unmounting is returning an error)
        // NewRun page also does not create a snackbar when create is clicked
        // this.props.updateSnackbar({
        //   message: `Successfully updated Run: ${updatedRun.name}`,
        //   open: true,
        // });
      } catch (err) {
        const errorMessage = await errorToMessage(err);
        this.showErrorDialog('Update failed', errorMessage);
        logger.error('Error updating Run:', err);
        return;
      } finally {
        this.setStateSafe({ isSaving: false });
      }
    });
  }
}

export default RecurringRunEdit;
