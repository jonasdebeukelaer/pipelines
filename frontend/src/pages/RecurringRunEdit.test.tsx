import { ApiJob } from '../apis/job';
import { ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { Apis } from '../lib/Apis';
import { PageProps } from './Page';
import { RoutePage, RouteParams } from '../components/Router';
import TestUtils from '../TestUtils';
import RecurringRunDetails from './RecurringRunDetails';
import * as React from 'react';
import RecurringRunEdit from './RecurringRunEdit';
import { ButtonKeys } from '../lib/Buttons';

describe('EditRecurringRun', () => {
  let tree: ReactWrapper<any> | ShallowWrapper<any>;
  let fullTestJob: ApiJob = {};

  const updateBannerSpy = jest.fn();
  const updateDialogSpy = jest.fn();
  const updateSnackbarSpy = jest.fn();
  const updateToolbarSpy = jest.fn();
  const historyPushSpy = jest.fn();
  const getJobSpy = jest.spyOn(Apis.jobServiceApi, 'getJob');
  const updateJobSpy = jest.spyOn(Apis.jobServiceApi, 'updateJob');

  function generateProps(): PageProps {
    const match = {
      isExact: true,
      params: { [RouteParams.runId]: fullTestJob.id },
      path: '',
      url: '',
    };
    return TestUtils.generatePageProps(
      RecurringRunDetails,
      '' as any,
      match,
      historyPushSpy,
      updateBannerSpy,
      updateDialogSpy,
      updateToolbarSpy,
      updateSnackbarSpy,
    );
  }

  beforeEach(() => {
    fullTestJob = {
      created_at: new Date(2018, 8, 5, 4, 3, 2),
      description: 'test job description',
      enabled: true,
      id: 'test-job-id',
      max_concurrency: '50',
      no_catchup: true,
      name: 'test job',
      pipeline_spec: {
        parameters: [{ name: 'param1', value: 'value1' }],
        pipeline_id: 'some-pipeline-id',
      },
      trigger: {
        periodic_schedule: {
          end_time: new Date(2018, 10, 9, 8, 7, 6),
          interval_second: '3600',
          start_time: new Date(2018, 9, 8, 7, 6),
        },
      },
    } as ApiJob;

    jest.clearAllMocks();
    getJobSpy.mockImplementation(() => fullTestJob);
    updateJobSpy.mockImplementation(() => undefined);
  });

  afterEach(() => tree.unmount());

  it('renders the page with details filled in', async () => {
    tree = shallow(<RecurringRunEdit {...generateProps()} />);
    await TestUtils.flushPromises();
    expect(tree).toMatchSnapshot();
  });

  it('clears the banner when load is called', async () => {
    tree = shallow(<RecurringRunEdit {...(generateProps() as any)} />);
    expect(updateBannerSpy).toHaveBeenCalledTimes(1);
    (tree.instance() as RecurringRunEdit).load();
    await TestUtils.flushPromises();
    expect(updateBannerSpy).toHaveBeenCalledTimes(2);
    expect(updateBannerSpy).toHaveBeenLastCalledWith({});
  });

  it('has a refresh and delete button in the toolbar', () => {
    tree = shallow(<RecurringRunDetails {...generateProps()} />);
    const instance = tree.instance() as RecurringRunDetails;

    const refreshBtn = instance.getInitialToolbarState().actions[ButtonKeys.REFRESH];
    const deleteBtn = instance.getInitialToolbarState().actions[ButtonKeys.DELETE_RUN];

    expect(refreshBtn).toBeDefined();
    expect(deleteBtn).toBeDefined();
  });

  it('allows updating the recurring run name', async () => {
    tree = shallow(<RecurringRunEdit {...(generateProps() as any)} />);
    await TestUtils.flushPromises();

    tree
      .find('#runName')
      .at(0)
      .simulate('change', { target: { value: 'new run name' } });

    expect((tree.instance() as RecurringRunEdit).state.runName).toBe('new run name');
  });

  it('allows updating the description', async () => {
    tree = shallow(<RecurringRunEdit {...(generateProps() as any)} />);
    await TestUtils.flushPromises();

    tree
      .find('#description')
      .at(0)
      .simulate('change', { target: { value: 'new description' } });

    expect((tree.instance() as RecurringRunEdit).state.description).toBe('new description');
  });

  it('renders an error if required field is missing', async () => {
    tree = shallow(<RecurringRunEdit {...(generateProps() as any)} />);
    await TestUtils.flushPromises();

    tree
      .find('#runName')
      .at(0)
      .simulate('change', { target: { value: '' } });

    expect(
      tree
        .find('#errorMessage')
        .at(0)
        .text(),
    ).toBe('Run name is required');
  });

  it('renders a warning if a run param is missing', async () => {
    tree = shallow(<RecurringRunEdit {...(generateProps() as any)} />);
    await TestUtils.flushPromises();

    let p = (tree.instance() as RecurringRunEdit).state.parameters;
    p[0].value = '';
    (tree.instance() as RecurringRunEdit).setState({ parameters: p });

    expect(
      tree
        .find('#missingParametersMessage')
        .at(0)
        .text(),
    ).toBe('Some parameters are missing values');
  });

  it('exits to the recurring run details if cancels', async () => {
    tree = shallow(<RecurringRunEdit {...(generateProps() as any)} />);
    await TestUtils.flushPromises();

    tree.find('#cancelBtn').simulate('click');

    let id = (tree.instance() as RecurringRunEdit).state.run!.id!;
    let path = RoutePage.RECURRING_RUN.replace(':' + RouteParams.runId, id);
    expect(historyPushSpy).toHaveBeenCalledWith(path);
  });

  describe('clicks save', () => {
    it('save button is clickable', async () => {
      tree = shallow(<RecurringRunEdit {...(generateProps() as any)} />);
      await TestUtils.flushPromises();

      expect(
        tree
          .find('#saveBtn')
          .at(0)
          .prop('disabled'),
      ).toBeFalsy();
    });

    it('shows dialogue on error, re-enable save button', async () => {
      tree = shallow(<RecurringRunEdit {...(generateProps() as any)} />);
      await TestUtils.flushPromises();

      TestUtils.makeErrorResponseOnce(updateJobSpy, 'test error message');

      tree.find('#saveBtn').simulate('click');
      await TestUtils.flushPromises();

      expect(updateDialogSpy).toHaveBeenCalledTimes(1);
      expect(updateDialogSpy).toHaveBeenCalledWith({
        buttons: [{ text: 'Dismiss' }],
        content: 'test error message',
        title: 'Update failed',
      });
      expect((tree.instance() as RecurringRunEdit).state.isSaving).toBe(false);
    });

    it('sends update request', async () => {
      tree = shallow(<RecurringRunEdit {...(generateProps() as any)} />);
      await TestUtils.flushPromises();

      tree
        .find('#runName')
        .at(0)
        .simulate('change', { target: { value: 'new run name' } });
      tree.find('#saveBtn').simulate('click');

      let job = (tree.instance() as RecurringRunEdit).state.run;
      job!.description = 'new run name';
      expect(updateJobSpy).toHaveBeenCalledTimes(1);
      expect(updateJobSpy).toHaveBeenCalledWith(job);
    });

    it('exits to the recurring run details page', async () => {
      tree = shallow(<RecurringRunEdit {...(generateProps() as any)} />);
      await TestUtils.flushPromises();

      tree.find('#saveBtn').simulate('click');
      await TestUtils.flushPromises();

      const id = (tree.instance() as RecurringRunEdit).state.run!.id!;
      let path = RoutePage.RECURRING_RUN.replace(':' + RouteParams.runId, id);
      expect(historyPushSpy).toHaveBeenCalledTimes(1);
      expect(historyPushSpy).toHaveBeenCalledWith(path);
    });
  });
});
