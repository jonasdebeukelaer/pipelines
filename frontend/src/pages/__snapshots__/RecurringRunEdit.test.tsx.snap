// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditRecurringRun renders the page with details filled in 1`] = `
<div
  className="page"
>
  <div
    className="page"
  >
    <div
      className=""
    >
      <div
        className="header"
      >
        General Params
      </div>
      <Input
        error={false}
        id="runName"
        label="Run name"
        multiline={false}
        onChange={[Function]}
        required={true}
        value="test job"
        variant="outlined"
      />
      <Input
        id="description"
        label="Description"
        multiline={false}
        onChange={[Function]}
        value="test job description"
        variant="outlined"
      />
      <Input
        id="serviceAccount"
        label="Service Account"
        multiline={false}
        onChange={[Function]}
        value=""
        variant="outlined"
      />
      <div
        className="header"
      >
        Run trigger
      </div>
      <Trigger
        catchup={true}
        maxConcurrentRuns="50"
        onChange={[Function]}
        trigger={
          Object {
            "periodic_schedule": Object {
              "end_time": 2018-11-09T08:07:06.000Z,
              "interval_second": "3600",
              "start_time": 2018-10-08T06:06:00.000Z,
            },
          }
        }
      />
      <RunParameters
        handleParamChange={[Function]}
        initialParams={
          Array [
            Object {
              "name": "param1",
              "value": "value1",
            },
          ]
        }
        titleMessage="Edit pipeline parameters"
      />
    </div>
    <div
      className="flex"
    >
      <BusyButton
        busy={false}
        className="buttonAction"
        disabled={false}
        id="saveBtn"
        onClick={[Function]}
        title="Save"
      />
      <WithStyles(Button)
        id="cancelBtn"
        onClick={[Function]}
      >
        Cancel
      </WithStyles(Button)>
      <div
        className=""
        id="errorMessage"
        style={
          Object {
            "color": "red",
          }
        }
      />
    </div>
  </div>
</div>
`;
